Stepped wizard widget

maier49 opened the pull request:
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [x] For new widgets, an entry has been added to the `.dojorc`
* [x] For new widgets, `theme.variant()` is added to the root domnode
* [x] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [x] WidgetProperties are exported

<!--
Our bots should ensure:

* [ ] All contributors have signed a CLA
* [ ] The PR passes CI testing
* [ ] Code coverage is maintained
* [ ] The PR has been reviewed and approved
-->

**Description:**
Adds a stepped wizard widget with horizontal and vertical display modes
Resolves #1125

tomdye commented:
I've taken a look at the examples in parade and I think this could be made more compatible with custom elements by simply passing `<Step>` widgets that include the title/subtitle etc. I think that the `StepWizard` widget itself can handle the rendering / clicking etc of the avatars thus meanign there's no need to use a render function.

Something like

```tsx
<Wizard onStep={(index: number) => {}} steps={3} initalStep={1}>
	<Step>{
		title: 'foo',
		subtitle: <span>foo subtitle</span>
	}</Step>
	<Step>{
		title: 'bar',
		subtitle: <span>bar subtitle</span>
	}</Step>
</Wizard>
```

As per the speed-dial, the wizard can just iterate through it's children.
What do you think about renaming this to `Wizard` / `Step` also?

maier49 commented:
@tomdye The naming makes sense. I'll make those changes

maier49 commented:
@tomdye Updated 👍

Since we're rendering the avatar/icon ourselves I also made some changes to explicitly handle an error state as we're not passing it to a render function.

tomdye commented:
@maier49 can you disable the user interaction as well? ie. if the wizard steps are to be controlled by next / prev buttons or by a different process.
Also, I think the styling might need some work. It doesn't look quite right just yet with the text wrapping and no differentiation between title/subtitle etc.

maier49 commented:
@tomdye I added a `clickable` property that defaults to `true` that turns off the pointer styles and ignores clicks if set to `false`. I also made another pass and I think the text and icons should be much closer to the ant version in terms of weight/size/opacity.

maier49 commented:
@tomdye I've reworked the API, added some basic styling to the default theme, and commented the classes. Could you be any more specific on what styles need updates?

tomdye commented on src/examples/src/widgets/wizard/Basic.tsx:
```diff
@@ -0,0 +1,95 @@
+					icache.set('activeStep', stepIndex);
+					icache.set(
+						'steps',
+						steps.map((_, index) => {
```
won't this overwrite any existing `errror` states?

maier49 replied:
Yes, but there aren't any in this basic example

tomdye replied:
fair enough, just thinking about the best thing to be showing users is all, it's not a bit issue

maier49 replied:
Fair enough, I can add an if to return `error` if there is one. No harm in doing so

tomdye commented on src/wizard/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	middleware: { theme, dimensions, resize }
+}) {
+	const classes = theme.classes(css);
+	const { direction = 'horizontal', onStep, steps, clickable = true } = properties();
```
do you think clickable should be the default?

maier49 replied:
That's reasonable, I'll flip this

tomdye commented on src/wizard/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	const width = dimensions.get('root').size.width;
+
+	const stepNodes = children();
+	const forceVertical = width < 200 * stepNodes.length;
```
where does `200` come from here and what is the reason behind forcing it to be vertical?

maier49 replied:
Forcing it to be vertical is basically copying Ant design's approach to responsiveness, except they only do it when viewed on "mobile" (via emulator at least), otherwise the steps just become squished and ugly/unusable. 200 px is just about how much space a step with a label, subtitle, and description needs to look reasonable. Could allow that limit to be overridden with a property?

Also open to any other suggestions on making this properly responsive

maier49 replied:
I've removed all responsive rendering logic from the widget. For now this will be the responsibility of the parent

tomdye commented on src/wizard/index.tsx:
```diff
@@ -0,0 +1,133 @@
+				theme={theme.compose(
+					avatarCss,
+					css,
+					'stepAvatar'
```
could you make this a single word, perhaps just `avatar`

tomdye commented on src/wizard/index.tsx:
```diff
@@ -0,0 +1,133 @@
+				)}
+				outline={Boolean(steps[index].status !== 'inProgress')}
+			>
+				{steps[index].status === 'complete' ? (
```
could you make this a switch to avoid the double ternary?

tomdye commented on src/wizard/tests/unit/Wizard.spec.tsx:
```diff
@@ -0,0 +1,237 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+const { it, describe } = intern.getInterface('bdd');
+const { assert } = intern.getPlugin('chai');
+import testHarness from '@dojo/framework/testing/harness/harness';
```
Do you think you could try using the new test renderer?
docs: https://next.dojo.io/learn/testing/introduction#using-assertions

`import renderer, { assertion } from '@dojo/framework/testing/renderer';`

tomdye replied:
Ant recently re-wrote the `List` tests using the new test renderer, would be a good place to start looking: https://github.com/dojo/widgets/blob/master/src/list/tests/List.spec.tsx

tomdye commented on src/wizard/index.tsx:
```diff
@@ -0,0 +1,148 @@
+				content = String(index + 1);
+		}
+		return {
+			content: [
```
How about using `<virtual>` here instead?
